### PR TITLE
[PKD-42]: Migrate to express

### LINKED ISSUE #42: Migrate to express

### PR DESCRIPTION
**Pull request by almogko**
- using express
- - refactor folder structure

### CODE REVIEW

File: server/controllers/pokemons/index.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    if (num > 99) {
+        return `#${num}`
+    }
+
+    if (num > 9) {
+        return `#0${num}`
+    }
+
+    return `#00${num}`
+}
```
Comment (eidan66) [inline server/controllers/pokemons/index.ts]: Let's move this function to utils folder please

File: server/controllers/pokemons/index.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+function mapToAPIPokemonType(type: string): APIPokemonType {
```
Comment (eidan66) [inline server/controllers/pokemons/index.ts]: why the type is a string?

File: server/controllers/pokemons/index.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    const apiType = Object.keys(APIPokemonType).find((s) => s.toLowerCase() === type.toLowerCase());    
+
+    if (!apiType) {
+        return APIPokemonType.Unknown;
+    }
+    
+    return apiType as APIPokemonType;
+}
+
+function pokeAPIPokemonToAPIPokemon(pokeApiPokemon: Pokemon): APIPokemon {
```
Comment (eidan66) [inline server/controllers/pokemons/index.ts]: Please find another name for this function, is isn't clear enough

File: server/controllers/pokemons/index.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import { Pokemon, PokemonClient } from 'pokenode-ts';
```
Comment (eidan66) [inline server/controllers/pokemons/index.ts:1]: Please replace all functions to be an arrow function.

File: server/controllers/pokemons/index.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    };
+}
+
+async function getPokemonsPage(offset: number, limit: number): Promise<APIPageResponse<APIPokemon>> {
+    const pokemonsPage = await api.listPokemons(offset, limit);
+
+    const apiPokemonsPage: APIPageResponse<APIPokemon> = {
+        count: pokemonsPage.count,
+        next: pokemonsPage.next?.replace('https://pokeapi.co/api/v2/pokemon', 'localhost:3000/pokemons'),
+        previous: pokemonsPage.previous?.replace('https://pokeapi.co/api/v2/pokemon', 'localhost:3000/pokemons'),
```
Comment (eidan66) [inline server/controllers/pokemons/index.ts]: Don't we need the `http` when you replace the URI?
```suggestion
        next: pokemonsPage.next?.replace('https://pokeapi.co/api/v2/pokemon', 'http://localhost:3000/pokemons'),
        previous: pokemonsPage.previous?.replace('https://pokeapi.co/api/v2/pokemon', 'http://localhost:3000/pokemons'),
```

File: server/controllers/pokemons/index.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        results: await Promise.all(pokemonsPage.results.map(async pokemon => {
```
Comment (eidan66) [inline server/controllers/pokemons/index.ts]:
```suggestion
        results: await Promise.all(pokemonsPage.results.map(async ({name}) => {
```

File: server/controllers/pokemons/index.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            const item = await api.getPokemonByName(pokemon.name);
```
Comment (eidan66) [inline server/controllers/pokemons/index.ts]:
```suggestion
            const item = await api.getPokemonByName(name);
```

File: server/controllers/pokemons/index.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    return apiPokemonsPage;
+}
+
+async function getPokemonByName(name: string): Promise<APIPokemon> {
+    const pokemon = await api.getPokemonByName(name);
+    const apiPokemon = pokeAPIPokemonToAPIPokemon(pokemon);
+
+    return apiPokemon;
```
Comment (eidan66) [inline server/controllers/pokemons/index.ts]: Please find another name to APIPokemon, it isn't clear enough

File: server/models/pokemon/index.ts [hunk 9]
```diff
@@ -28,6 +21,27 @@ export enum APIPokemonType {
...
+    Grass: '#7AC74C',
+    Ground: '#E2BF65',
+    Ice: '#96D9D6',
+    Normal: '#A8A77A',
+    Poison: '#A33EA1',
+    Psychic: '#F95587',
+    Rock: '#B6A136',
+    Steel: '#B7B7CE',
+    Water: '#6390F0',
+  } as  { [key in APIPokemonType]: string };
```
Comment (eidan66) [inline server/models/pokemon/index.ts]:
```suggestion
  } as  Record<APIPokemonType,string>;
```

File: server/models/response/index.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+export interface APIPageResponse<T> {
+    count: number,
+    next:string | null | undefined,
+    previous: string | null | undefined,
```
Comment (eidan66) [inline server/models/response/index.ts]:
```suggestion
    next?:string | null,
    previous?: string | null, 
```

Comment (eidan66) [reply server/models/response/index.ts]: And why we holding `null` when the value can be undefined?

File: server/routes/index.ts [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+const routes  = express.Router();
+
+
+routes.use('/pokemons', pokemons);
+
+routes.use(function(req, res) {
+    res.statusCode = 404;
+    res.end();
+});
```
Comment (eidan66) [inline server/routes/index.ts]: Arrow function please

File: server/routes/pokemons/index.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+routes.route('/:name')
+  .get(async (req, res) => {
+    const name = req.params.name;
+    const pokemon = await pokemonController.getPokemonByName(name);    
+    res.end(JSON.stringify(pokemon));
+  });
+
+routes.route('/')
+.get(async (req, res) => {
```
Comment (eidan66) [inline server/routes/pokemons/index.ts]: Please use prettier to fix all the unnecessary/missing spaces

File: server/routes/pokemons/index.ts [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    const defaultOffset = 0;
+    const defaultLimit = 20;
+    let offset = defaultOffset;
+    let limit = defaultLimit;
```
Comment (eidan66) [inline server/routes/pokemons/index.ts]: Why not?
```suggestion
    let offset = 0
    let limit = 20
```

File: server/routes/pokemons/index.ts [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+
+    if (req.query.offset && typeof req.query.offset === 'string') {
```
Comment (eidan66) [inline server/routes/pokemons/index.ts]:
```suggestion
    if ( typeof req?.query?.offset === 'string') {
```

File: server/routes/pokemons/index.ts [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+         offset = parseInt(req.query.offset);
+    }
+
+    if (req.query.limit && typeof req.query.limit === 'string') {
```
Comment (eidan66) [inline server/routes/pokemons/index.ts]:
```suggestion
    if ( typeof req?.query?.limit === 'string') {
```

File: server/controllers/pokemons/index.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+function pokeAPIPokemonToAPIPokemon(pokeApiPokemon: Pokemon): APIPokemon {
+    return {
+        id: pokeApiPokemon.id,
+        name: pokeApiPokemon.name,
+        number: pad(pokeApiPokemon.id),
+        types: pokeApiPokemon.types.map(({type:{name}})=> mapToAPIPokemonType(name)),
+        boxBg: COLORS[mapToAPIPokemonType(pokeApiPokemon.types[0].type.name)],
+        svg: `https://github.com/eidan66/pokemon-api-sprites/blob/master/sprites/pokemon/other/showdown/${pokeApiPokemon.id}.gif?raw=true`
+    };
+}
```
Comment (eidan66) [inline server/controllers/pokemons/index.ts]:
```suggestion
function pokeAPIPokemonToAPIPokemon({id,name,types}: Pokemon): APIPokemon {
    return {
        id,
        name,
        number: pad(id),
        types: types.map(({type:{name}})=> mapToAPIPokemonType(name)),
        boxBg: COLORS[mapToAPIPokemonType(types[0].type.name)],
        svg: `https://github.com/eidan66/pokemon-api-sprites/blob/master/sprites/pokemon/other/showdown/${pokeApiPokemon.id}.gif?raw=true`
    };
}
```

**Comment by eidan66** [comment]
General review:
1. Please use prettier to verify all the linter rules are applied.
2. Please use Arrow function when you create new functions.

File: server/controllers/pokemons/index.ts (hunk 2 above)
Comment (almogko) [reply server/controllers/pokemons/index.ts]: raw string from pokenode api

File: server/routes/pokemons/index.ts (hunk 13 above)
Comment (almogko) [reply server/routes/pokemons/index.ts]: just a standard im used to

File: server/routes/index.ts (hunk 11 above)
Comment (almogko) [reply server/routes/index.ts]: changed

File: server/controllers/pokemons/index.ts (hunk 8 above)
Comment (almogko) [reply server/controllers/pokemons/index.ts]: changed

File: server/controllers/pokemons/index.ts (hunk 7 above)
Comment (almogko) [reply server/controllers/pokemons/index.ts]: changed

File: server/controllers/pokemons/index.ts (hunk 4 above)
Comment (almogko) [reply server/controllers/pokemons/index.ts:1]: changed

File: server/controllers/pokemons/index.ts (hunk 3 above)
Comment (almogko) [reply server/controllers/pokemons/index.ts]: changed

File: server/controllers/pokemons/index.ts (hunk 1 above)
Comment (almogko) [reply server/controllers/pokemons/index.ts]: changed

File: server/routes/pokemons/index.ts (hunk 12 above)
Comment (almogko) [reply server/routes/pokemons/index.ts]: added

File: server/controllers/pokemons/index.ts [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+const mapPokeAPIPokemonToPokemonModel = ({ id, name, types }: Pokemon): PokemonModel => {
+  return {
+    id,
+    name,
+    number: Utils.pad(id),
+    types: types.map(({ type: { name } }) => mapToPokemonTypeModel(name)),
+    boxBg: COLORS[mapToPokemonTypeModel(types[0].type.name)],
+    svg: `https://github.com/eidan66/pokemon-api-sprites/blob/master/sprites/pokemon/other/showdown/${id}.gif?raw=true`,
+  };
+};
```
Comment (eidan66) [inline server/controllers/pokemons/index.ts]:
```suggestion
const mapPokeAPIPokemonToPokemonModel = ({ id, name, types }: Pokemon): PokemonModel => ({
    id,
    name,
    number: Utils.pad(id),
    types: types.map(({ type: { name } }) => mapToPokemonTypeModel(name)),
    boxBg: COLORS[mapToPokemonTypeModel(types[0].type.name)],
    svg: `https://github.com/eidan66/pokemon-api-sprites/blob/master/sprites/pokemon/other/showdown/${id}.gif?raw=true`,
  });
```

File: server/utils/index.ts [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+  }
+
+  if (num > 9) {
+    return `#0${num}`;
+  }
+
+  return `#00${num}`;
+};
+
+export default { pad };
```
Comment (eidan66) [inline server/utils/index.ts]:
```suggestion
export const pad = (num: number) => {
  if (num > 99) {
    return `#${num}`;
  }

  if (num > 9) {
    return `#0${num}`;
  }

  return `#00${num}`;
};
```

**Pull request merged**